MySensors config flow

functionpointer opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
MySensors integration is now configured via the frontend. Existing entries in `configuration.yaml` are automatically migrated. No changes are applied after the migration has happened. If you need to change anything, use the frontend.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Adds config flow support to MySensors.  
Adds unique IDs to all entities created by MySensors.  
Adds device information to all entities created by MySensors. MySensors nodes may be prepresented my multiple entities. With the device info every node will be represented as a device.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16270

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

functionpointer committed:
MySensors: Add type annotations

Adds a bunch of type annotations that were created
while understanding the code.

functionpointer committed:
MySensors: Change GatewayId to string

In preparation for config flow.
The GatewayId used to be id(gateway).

With config flows, every gateway will have its own
ConfigEntry. Every ConfigEntry has a unique id.
Thus we would have two separate but one-to-one related ID systems.

This commit removes this unneeded duplication by using the id of the ConfigEntry
as GatewayId.

functionpointer committed:
MySensors: Add unique_id to all entities

This allows entities to work well with the frontend.

functionpointer committed:
MySensors: Add device_info to all entities

Entities belonging to the same node_id will now by grouped as a device.

functionpointer committed:
MySensors: clean up device.py a bit

functionpointer committed:
MySensors: Add config flow support

With this change the MySensors can be fully configured from the GUI.

Legacy configuration.yaml configs will be migrated by reading them once.
Note that custom node names are not migrated. Users will have to re-enter
the names in the front-end.
Since there is no straight-forward way to configure global settings,
all previously global settings are now per-gateway. These settings include:
- MQTT retain
- optimistic
- persistence enable
- MySensors version

When a MySensors integration is loaded, it works as follows:
1. __init__.async_setup_entry is called
2. for every platform, async_forward_entry_setup is called
3. the platform's async_setup_entry is called
4. __init__.setup_mysensors_platform is called
5. the entity's constructor (e.g. MySensorsCover) is called
6. the created entity is stored in a dict in the hass object

functionpointer committed:
MySensors: Fix linter errors

homeassistant commented:
Hi @functionpointer, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#45421).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,176 @@
+                )
+
+        schema = OrderedDict()
+        schema[vol.Optional(CONF_OPTIMISTIC, default=False)] = bool
```
I want to deprecate optimistic. We request echo (ack) for values sent so there shouldn't be a need anymore.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,176 @@
+
+        schema = OrderedDict()
+        schema[vol.Optional(CONF_OPTIMISTIC, default=False)] = bool
+        schema[vol.Optional(CONF_PERSISTENCE, default=True)] = bool
```
I want to deprecate persistence option and always use persistence.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,176 @@
+        schema = OrderedDict()
+        schema[vol.Optional(CONF_OPTIMISTIC, default=False)] = bool
+        schema[vol.Optional(CONF_PERSISTENCE, default=True)] = bool
+        schema[vol.Optional(CONF_RETAIN, default=True)] = bool
```
This options should only be shown for MQTT gateways. Do a two step form flow where first selecting gateway type and then the gateway options required to connect.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,176 @@
+        schema[vol.Optional(CONF_VERSION, default=DEFAULT_VERSION)] = str
+
+        schema[vol.Required(CONF_DEVICE, default="127.0.0.1")] = str
+        # schema[vol.Optional(CONF_PERSISTENCE_FILE)] = str
```
```suggestion
```

functionpointer replied:
The option is currently used for importing from `configuration.yaml`. The comment serves as a hint that it is still there and could be activated in the gui.

I see several options:
- Get rid of the comment only, keep using the option
- Get rid of the option and have the migration code rename the file to the id
- Get rid of the option and don't rename the file. This means we won't load existing persistence files after migration

MartinHjelmare replied:
We won't have this option in the form. Imported items won't go via the form validation.

We may want to separate the import items processing from the user flow since we want a two step flow for the user, where the first step selects gateway type. That won't work for the imported data directly.

MartinHjelmare replied:
We should have some migration code to allow persistence file to be migrated.

functionpointer replied:
I have chosen option 1.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,176 @@
+        schema[vol.Required(CONF_DEVICE, default="127.0.0.1")] = str
+        # schema[vol.Optional(CONF_PERSISTENCE_FILE)] = str
+        schema[
+            vol.Optional(CONF_BAUD_RATE, default=DEFAULT_BAUD_RATE)
```
This is only for serial gateways.

functionpointer committed:
MySensors: Remove unused import

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,176 @@
+        schema[
+            vol.Optional(CONF_BAUD_RATE, default=DEFAULT_BAUD_RATE)
+        ] = cv.positive_int
+        schema[vol.Optional(CONF_TCP_PORT, default=DEFAULT_TCP_PORT)] = int
```
Only for tcp gateway.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,176 @@
+            vol.Optional(CONF_BAUD_RATE, default=DEFAULT_BAUD_RATE)
+        ] = cv.positive_int
+        schema[vol.Optional(CONF_TCP_PORT, default=DEFAULT_TCP_PORT)] = int
+        schema[vol.Optional(CONF_TOPIC_IN_PREFIX)] = str
+        schema[vol.Optional(CONF_TOPIC_OUT_PREFIX)] = str
```
Only for MQTT gateway.

MartinHjelmare commented on homeassistant/components/mysensors/const.py:
```diff
@@ -1,33 +1,55 @@
+CHILD_CALLBACK: str = "mysensors_child_callback_{}_{}_{}_{}"
+NODE_CALLBACK: str = "mysensors_node_callback_{}_{}"
+MYSENSORS_DISCOVERY = "mysensors_discovery_{}_{}"
+MYSENSORS_ON_UNLOAD = "MYSENSORS_ON_UNLOAD_{}"
```
Please use lowercase snake_case strings.

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -155,67 +166,79 @@ def internal_callback(msg):
     """Start the gateway."""
     # Don't use hass.async_create_task to avoid holding up setup indefinitely.
-    connect_task = hass.loop.create_task(gateway.start())
+    gateway.connect_task = hass.loop.create_task(gateway.start())
```
Why are we storing the task as an attribute on the gateway?
```suggestion
    gateway.connect_task = asyncio.create_task(gateway.start())
```

functionpointer replied:
It is stored so that it can be cancelled in `gw_stop()`. I will add a comment to document that.

MartinHjelmare replied:
I'd prefer to store it in hass.data so we don't extend the attributes on the Gateway further. That's a bad pattern already for the other things we store the same way.

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -155,67 +166,79 @@ def internal_callback(msg):
-
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, gw_stop)
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP, lambda event: gw_stop(hass, gateway)
```
Lambdas will run in the executor. That doesn't work with functions that should run in the event loop.

MartinHjelmare commented on homeassistant/components/mysensors/manifest.json:
```diff
@@ -2,7 +2,14 @@
+    "@MartinHjelmare", "@functionpointer"
+  ],
+  "config_flow": true
+}
```
```suggestion
}
```

MartinHjelmare commented on homeassistant/components/mysensors/strings.json:
```diff
@@ -0,0 +1,14 @@
+  "title": "MySensors",
+  "config": {
+    "step": {
+      "confirm": {
```
We need to add strings for the form data.

functionpointer replied:
Which form data?
Also, i now have created a dropdown with `vol.In("a","b")`. Is there a way to add strings for the options to the translations?

MartinHjelmare replied:
The items we ask for input from the user in the config flow forms. Each item needs a string here. See the docs:
https://developers.home-assistant.io/docs/internationalization/core#config--options

MartinHjelmare replied:
Items in a `vol.In` validator can not be translated. If we use the gateway type names here (Serial, TCP, MQTT) the need for translation is minimized.

MartinHjelmare commented on homeassistant/components/mysensors/switch.py:
```diff
@@ -32,43 +47,56 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+            for device in _devices:
+                await device.async_turn_on(**kwargs)
+
+        hass.services.async_register(
```
In a future PR we should change to use the entity service helper:
https://developers.home-assistant.io/docs/dev_101_services#entity-services

MartinHjelmare replied:
We should only register the service once, in the scope of `async_setup_entry`.

MartinHjelmare reviewed (changes requested):
Thanks for the PR!

Most things look good.

I'm planning to deprecate many of the options that we currently support so I don't want to include support for those in the config flow forms. I think it's good to do that deprecation now when we're making a breaking change anyway.

We can't rely on config details to generate a unique id. Let's use the config entry id as unique gateway id for the entity unique id and skip a unique id for config entries themselves.

functionpointer commented on homeassistant/components/mysensors/__init__.py:
```diff
@@ -34,7 +44,7 @@
 DEFAULT_BAUD_RATE = 115200
 DEFAULT_TCP_PORT = 5003
-DEFAULT_VERSION = "1.4"
+DEFAULT_VERSION = "2.3"
```
I am not sure either, but don't you think it is unlikely that a new deployment would be using an old version?
Maybe we should not have a default at all and force the user to enter the version. An empty free-text field might be confusing for the user though. Maybe we should have no default value and 2.3 as suggested value?

I fully agree auto-detect would be amazing.

MartinHjelmare replied:
I'm thinking that 1.4 may be the safest version to default to since it won't activate features that require a higher version.

I'll like the suggestion with requiring input version and have a suggested value. :+1:

MartinHjelmare commented:
When we've settled on how to structure the config flow, don't forget to add tests for the config flow. It requires 100 % test coverage. Please use our scaffold script template to generate best practice code for the tests.

functionpointer commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,176 @@
+                )
+            elif not is_serial:
+                uniquestr += str(user_input[CONF_TCP_PORT])
+            gateway_id = hashlib.sha256(uniquestr.encode()).hexdigest()[:8]
```
what "entry id" do you mean? hopefully not `id()` of the ConfigEntry?

MartinHjelmare replied:
The `ConfigEntry.entry_id` that is available after the entry is created.

functionpointer replied:
Okay, but what about persistence file naming? The previous scheme allowed the user to delete and re-create an entry and it would load the already existing persistence file. I found that quite useful, altough maybe that was just because i was debugging config flows

MartinHjelmare replied:
Yeah, that's a good point. I think we can keep the persistence file option, but make it optional. If it doesn't exist we generate a default name, eg based on the config entry entry_id in the integration setup.

functionpointer replied:
Ok, i have added it. Ideally i would like to hide it behind some sort of advanced toggle, but i dont think that is supported in config flows.

functionpointer committed:
MySensors: Feedback from @MartinHjelmare

MartinHjelmare commented on homeassistant/components/mysensors/__init__.py:
```diff
@@ -110,8 +108,6 @@ def validator(config):
                     vol.Required(CONF_GATEWAYS): vol.All(
                         cv.ensure_list, has_all_unique_files, [GATEWAY_SCHEMA]
                     ),
-                    vol.Optional(CONF_OPTIMISTIC, default=False): cv.boolean,
```
Wts keep these validators but also add `cv.deprecated` to mark the items as deprecated.

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -217,22 +217,26 @@ async def _discover_persistent_devices(
     hass.bus.async_listen_once(
-        EVENT_HOMEASSISTANT_STOP, lambda event: gw_stop(hass, gateway)
+        EVENT_HOMEASSISTANT_STOP,
+        lambda event: asyncio.create_task(gw_stop(hass, gateway)),
```
I think we can have a coroutine function as event callback, which awaits the gateway stop function.

functionpointer committed:
MySensors: Multi-step config flow

functionpointer committed:
MySensors: More feedback

functionpointer committed:
MySensors: Move all storage in hass object under DOMAIN

The integration now stores everything under hass.data["mysensors"]
instead of using several top level keys.

functionpointer committed:
MySensors: await shutdown of gateway instead of creating a task

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -155,68 +167,84 @@ def internal_callback(msg):
-        hass.async_create_task(gateway.stop())
-        if not connect_task.done():
-            connect_task.cancel()
+    async def stop_this_gw():
```
We need to accept an argument which is a core event.
```suggestion
    async def stop_this_gw(_: Event):
```

functionpointer replied:
Ah okay i will add that.
Is there no documentation about this or am i just incapable of using a search engine?

MartinHjelmare commented on homeassistant/components/mysensors/__init__.py:
```diff
@@ -11,13 +11,15 @@
 
 from ... import config_entries
 from ...config_entries import ConfigEntry
+from ...const import CONF_OPTIMISTIC
```
Please use absolute imports for the non mysensors integration package imports. Intra package imports should still be relative.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -76,101 +83,208 @@ def gateway_ready_callback(msg):
 
+def _get_schema_common() -> dict:
+    """Create a schema with options common to all gateway types."""
+    schema = dict()
```
We can create the dict and populate it in one go.
```py
schema = {
    vol.Required(
        CONF_VERSION, default="", description={"suggested_value": DEFAULT_VERSION}
    ): str
}

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -76,101 +83,208 @@ def gateway_ready_callback(msg):
+
+        if (
+            user_input is not None
+            and user_input.get(CONF_GATEWAY_TYPE) in CONF_GATEWAY_TYPE_ALL
```
This is validated by the form schema.

MartinHjelmare commented on homeassistant/components/mysensors/const.py:
```diff
@@ -1,19 +1,33 @@
 CONF_VERSION: str = "version"
+CONF_GATEWAY_TYPE: str = "gateway_type"
+CONF_GATEWAY_TYPE_TYPE = Literal[
+    "gateway_type_serial", "gateway_type_ethernet", "gateway_type_mqtt"
```
Technically we only support the TCP gateway type. MySensors have an option for UDP which we don't support for ethernet gateway.

functionpointer committed:
MySensors: Rename Ethernet to TCP

functionpointer committed:
MySensors: Absolute imports and cosmetic changes

functionpointer committed:
MySensors: fix gw_stop

functionpointer committed:
MySensors: Allow user to specify persistence file

functionpointer committed:
MySensors: Nicer log message

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -96,9 +96,9 @@ async def _get_gateway(
-    persistence_file = data.get(
-        CONF_PERSISTENCE_FILE, hass.config.path(f"mysensors_{unique_id}.pickle")
-    )
+    persistence_file = data.get(CONF_PERSISTENCE_FILE, f"mysensors_{unique_id}.pickle")
```
It should be json not pickle.

MartinHjelmare commented on .coveragerc:
```diff
@@ -568,7 +568,14 @@ omit =
     homeassistant/components/mycroft/notify.py
-    homeassistant/components/mysensors/*
+    homeassistant/components/mysensors/binary_sensor.py
+    homeassistant/components/mysensors/light.py
```
Please exclude all modules except the config flow module.

functionpointer committed:
MySensors: Add lots of unit tests

functionpointer committed:
MySensors: Fix legacy import of persistence file name

Turns out tests help to find bugs :D

functionpointer committed:
MySensors: Improve test coverage

functionpointer committed:
MySensors: Use json persistence files by default

functionpointer commented:
@MartinHjelmare i have run into a bit of a roadblock with all the linter madness:

`Warning: homeassistant/components/mysensors/config_flow.py:18:0: W0611: Unused DOMAIN imported from const (unused-import)`

I am importing `DOMAIN` because it is used in the definition of the config flow handler class `MySensorsConfigFlowHandler`, as recommended by the [documentation](https://developers.home-assistant.io/docs/config_entries_config_flow_handler).

However, pylint apparently does not recognize that as a usage and recommends to remove the import.
I have tried to silence the warning with `# pylint: disable=unused-import` as recommended by the [documentation](https://developers.home-assistant.io/docs/development_testing/#notes-on-pylint-and-pep8-validation), but the comment only works when applied to the entire import statement. This would end up ignoring 14 imports from `.const` instead of just one.

What is the best course of action here?

MartinHjelmare commented:
Disable the pylint warning locally as the docs suggests.

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -58,48 +60,103 @@ def is_socket_address(value):
-async def _get_gateway(hass, config, gateway_conf, persistence_file):
+async def _get_gateway(
+    hass: HomeAssistantType,
+    entry: Union[ConfigEntry, Dict[str, Any]],
```
We should consolidate the signature to only accept a config entry and not a union.

functionpointer replied:
The dict variant is used by `try_connect()` which is executed before a config entry is ready.
How should i attempt to connect without the dict variant?

MartinHjelmare replied:
Just pass the needed gateway parameters. It's better to decrease the scope of each function and not require the function to know about things not really related to its task.

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -58,48 +60,103 @@ def is_socket_address(value):
+    user_input_copy = user_input.copy()
+    try:
+        gateway: Optional[BaseAsyncGateway] = await _get_gateway(
+            hass, user_input_copy, str(randint(0, 10 ** 6)), persistence=False
```
We shouldn't generate an id with randint.

functionpointer replied:
This id is very short lived (only in `try_connect()`) and does not get stored anywhere.

What should i use instead?

MartinHjelmare replied:
The gateway doesn't require a random string to be created.

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -155,68 +212,84 @@ def internal_callback(msg):
     gateway.device = device
-    gateway.event_callback = _gw_callback_factory(hass, config)
-    gateway.nodes_config = gateway_conf[CONF_NODES]
+    gateway.entry_id = unique_id
```
We shouldn't set additional attributes on the gateway.

MartinHjelmare commented on homeassistant/components/mysensors/device_tracker.py:
```diff
@@ -18,17 +23,25 @@ async def async_setup_scanner(hass, config, async_see, discovery_info=None):
-        gateway_id = id(device.gateway)
-        dev_id = (gateway_id, device.node_id, device.child_id, device.value_type)
-        async_dispatcher_connect(
+        gateway_id: GatewayId = device.gateway.entry_id
```
We can pass the entry_id as part of the discovery info.

MartinHjelmare commented on homeassistant/components/mysensors/const.py:
```diff
@@ -107,14 +143,25 @@
     "sensor": SENSOR_TYPES,
     "switch": SWITCH_TYPES,
 }
+"""dict mapping hass platform name to dict that maps mysensors s_type to set of mysensors v_type"""
```
Comments are done with `# `.

functionpointer replied:
Those are supposed to be docstrings for these constants.
I know [PEP-257](https://www.python.org/dev/peps/pep-0257/#id15) says they are not "real docstrings" in the sense that they will be available as `__doc__`, but they are are allowed and considered "additional docstrings". And there are tools that read them: For example the PyCham IDE displays them whenever you hover over any usage of the constant.

Are you sure you want me to convert them to normal comments?

MartinHjelmare replied:
Yes, please. Personally I don't think we need this many comments.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,229 @@
+def _validate_version(version: str) -> Dict[str, str]:
+    """Validate a version string from the user."""
+    errors = {CONF_VERSION: "invalid_version"}
+    version_parts = version.split(".")
```
Can we parse the version with a version parser from eg `packaging` instead?

functionpointer replied:
Sounds like a good idea. In particular i would love to use `semver`. However, MySensors versions are not compliant to it, so we have to find something else.
The pymysensors library uses
```
# pylint: disable=no-name-in-module, import-error
from distutils.version import LooseVersion as parse_ver
```
I tried it out and found that LooseVersion really accepts anything. But there is also StrictVersion, which seems to work just fine.

Maybe we could use that? It would save us from adding another dependency.

MartinHjelmare replied:
Please use this:
```py
from packaging.version import parse as parse_version
```

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,229 @@
+
+    async def async_step_user(self, user_input: Optional[Dict[str, str]] = None):
+        """Create a config entry from frontend user input."""
+        schema = dict()
```
Please create dicts with the literal expression `{}`. We can populate the dict directly.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,229 @@
+        schema[vol.Required(CONF_GATEWAY_TYPE)] = vol.In(CONF_GATEWAY_TYPE_ALL)
+        schema = vol.Schema(schema)
+
+        if user_input is not None and CONF_GATEWAY_TYPE in user_input:
```
```suggestion
        if user_input is not None:
```

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,229 @@
+        entered from the frontend.
+        Therefore we process it as though it came from the frontend.
+        """
+        if user_input.get(CONF_DEVICE) == MQTT_COMPONENT:
```
```suggestion
        if user_input[CONF_DEVICE] == MQTT_COMPONENT:
```

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,229 @@
+
+import voluptuous as vol
+
+from homeassistant.components.mysensors import (
```
Intra package imports should be relative.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,229 @@
+import homeassistant.helpers.config_validation as cv
+
+from . import CONF_RETAIN, CONF_VERSION, DEFAULT_VERSION
+from ... import config_entries
```
Non intra package imports should be absolute.

functionpointer committed:
MySensors: Stop adding attributes to existing objects

This commit removes the extra attributes that were being
added to the gateway objects from pymysensors.

Most attributes were easy to remove, except for the gateway id.
The MySensorsDevice class needs the gateway id as it is part of its DevId
as well as the unique_id and device_info.
Most MySensorsDevices actually end up being Entities.
Entities have access to their ConfigEntry via self.platform.config_entry.

However, the device_tracker platform does not become an Entity.
For this reason, the gateway id is not fetched from self.plaform but
given as an argument.

Additionally, MySensorsDevices expose the address of the gateway
(CONF_DEVICE). Entities can easily fetch this information via self.platform,
but the device_tracker cannot. This commit chooses to remove the gateway
address from device_tracker. While this could in theory break some automations,
the simplicity of this solution was deemed worth it.
The alternative of adding the entire ConfigEntry as an argument to MySensorsDevices
is not viable, because device_tracker is initialized by the async_setup_scanner function
that isn't supplied a ConfigEntry. It only gets discovery_info.
Adding the entire ConfigEntry doesn't seem appropriate for this edge case.

functionpointer committed:
MySensors: Fix gw_stop and the translations

functionpointer commented on homeassistant/components/mysensors/__init__.py:
```diff
@@ -112,69 +125,169 @@ def validator(config):
-        node_name,
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Remove an instance of the MySensors integration."""
+    _LOGGER.debug("Unload entry: %s (id: %s)", entry.title, entry.entry_id)
```
Good point. I have removed several of them. Hopefully i managed to catch them all

homeassistant commented:
Hello @functionpointer,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`8ba929c28af69b1b2cc930fac9f1e2146585a07a`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;45421&#x2F;commits&#x2F;8ba929c28af69b1b2cc930fac9f1e2146585a07a) - This commit has something that looks like an email address (suspendfunction@googlemail.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

functionpointer committed:
MySensors: Code style improvements

functionpointer committed:
MySensors: Fix incorrect function calls

functionpointer committed:
MySensors: Fewer comments in const.py

functionpointer committed:
MySensors: Remove union from _get_gateway and remove id from try_connect

functionpointer committed:
MySensors: Deprecate nodes option in configuration.yaml

functionpointer committed:
MySensors: Use version parser from packaging

MartinHjelmare commented on homeassistant/components/mysensors/device.py:
```diff
@@ -19,33 +32,99 @@
+    def device_info(self) -> Optional[Dict[str, Any]]:
+        """Return a dict that allows home assistant to puzzle all entities belonging to a node together."""
+        return {
+            "identifiers": {(DOMAIN, f"mys{self.gateway_id}-{self.node_id}")},
```
We don't need the mys prefix before gateway id.

MartinHjelmare commented on homeassistant/components/mysensors/device.py:
```diff
@@ -19,33 +32,99 @@
+            "identifiers": {(DOMAIN, f"mys{self.gateway_id}-{self.node_id}")},
+            "name": self.node_name,
+            "manufacturer": DOMAIN,
+            "model": self.node_name,
```
If model is the same as name we don't need to add it. It's not a real model name.

MartinHjelmare commented on homeassistant/components/mysensors/device.py:
```diff
@@ -19,33 +32,99 @@
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID for use in home assistant."""
+        return f"mys{self.gateway_id}-{self.node_id}-{self.child_id}-{self.value_type}"
```
Remove the mys prefix.

MartinHjelmare commented on homeassistant/components/mysensors/device.py:
```diff
@@ -19,33 +32,99 @@
+                    )
+
+    @property
+    def _mysensors_sensor(self) -> Sensor:
```
This is a node. Please rename it appropriately.
```suggestion
    def _node(self) -> Sensor:
```

MartinHjelmare commented on homeassistant/components/mysensors/device.py:
```diff
@@ -19,33 +32,99 @@
+        return self.gateway.sensors[self.node_id]
+
+    @property
+    def _mysensors_childsensor(self) -> ChildSensor:
```
This is a child.
```suggestion
    def _child(self) -> ChildSensor:
```

MartinHjelmare commented on homeassistant/components/mysensors/device.py:
```diff
@@ -19,33 +32,99 @@
+            if platform_str in self.hass.data[DOMAIN]:
+                platform_dict = self.hass.data[DOMAIN][platform_str]
+                if self.dev_id in platform_dict:
+                    if platform_dict[self.dev_id] is not self:
```
When can this happen?

functionpointer replied:
It shouldn't. If it did, that would be a bug.

A bug that triggers this would mean a serious fault in the DevId -> MySensorsDevice mapping, which would probably manifest in lots of weird behaviour because messages would be delivered to the wrong entities. The check is quite cheap to run and would help immensely in such a situation.

I would recommend to keep it in, but of course if you don't like it i will remove it.

MartinHjelmare replied:
Bump.

MartinHjelmare commented on homeassistant/components/mysensors/device.py:
```diff
@@ -19,33 +32,99 @@
-    def __init__(self, gateway, node_id, child_id, name, value_type):
+    def __init__(
+        self,
+        gateway_id: GatewayId,
```
Isn't the gateway id just a string?

functionpointer replied:
It used to be `int` when the id was literally `id(gateway)`. I changed it to `str`.

The special type allows checkers to find some simple bugs and serves as a reminder of what you are dealing with.
The type is especially helpful when nested in tuples or dicts.

The type `DevId` was also added for the same reason.

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -58,48 +59,113 @@ def is_socket_address(value):
+
+async def _get_gateway(
+    hass: HomeAssistantType,
+    gateway_id: Optional[GatewayId],
```
Pass in the event callback and persistence file instead of gateway id.

functionpointer committed:
MySensors: Remove prefix from unique_id and change some private property names

functionpointer committed:
MySensors: Change _get_gateway function signature

functionpointer committed:
MySensors: add packaging==20.8 for the version parser

MartinHjelmare commented on homeassistant/components/mysensors/manifest.json:
```diff
@@ -3,7 +3,8 @@
   "requirements": [
-    "pymysensors==0.20.1"
+    "pymysensors==0.20.1",
+    "packaging==20.8"
```
I don't think we need this since we require `setuptools` to install home assistant and `setuptools` requires `packaging`.

functionpointer replied:
It was inconsistent for me. In one of my venvs it was there already but in the other it was missing. After i have added the requirement it worked.
But if you think it isn't needed then i will remove it.

MartinHjelmare replied:
`setuptools` has vendored `packaging` for three years. If the venv was created longer than three years ago it may not be included.

We can keep it for now but probably we'll switch to `awesomeversion` instead. If that becomes a core requirement we don't need a custom requirement here.

MartinHjelmare replied:
Please add `packaging` back to the manifest for now. We can clean it up later when we switch to `awesomeversion`.

MartinHjelmare replied:
We just added `awesomeversion` as a core requirement in Home Assistant. So now we can use that instead. Please change the code to use that package and remove the requirement on `packaging`here.

https://pypi.org/project/awesomeversion/

MartinHjelmare commented on homeassistant/components/mysensors/device.py:
```diff
@@ -19,33 +32,98 @@
+                    )
+
+    @property
+    def _mysensors_node(self) -> Sensor:
```
```suggestion
    def _node(self) -> Sensor:
```

MartinHjelmare commented on homeassistant/components/mysensors/device.py:
```diff
@@ -19,33 +32,98 @@
+        return self.gateway.sensors[self.node_id]
+
+    @property
+    def _mysensors_child(self) -> ChildSensor:
```
```suggestion
    def _child(self) -> ChildSensor:
```

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
 
-async def _discover_persistent_devices(hass, hass_config, gateway):
+async def _discover_persistent_devices(
+    hass: HomeAssistantType, hass_config: ConfigEntry, gateway: BaseAsyncGateway
```
```suggestion
    hass: HomeAssistantType, entry: ConfigEntry, gateway: BaseAsyncGateway
```

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
-            validated = validate_child(gateway, node_id, child)
+        node: Sensor = gateway.sensors[node_id]
+        for child in node.children.values():  # child is of type ChildSensor
+            validated = validate_child(hass_config.entry_id, gateway, node_id, child)
```
```suggestion
            validated = validate_child(entry.entry_id, gateway, node_id, child)
```

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
+    _LOGGER.debug("discovering persistent devices: %s", new_devices)
     for platform, dev_ids in new_devices.items():
-        tasks.append(discover_mysensors_platform(hass, hass_config, platform, dev_ids))
+        discover_mysensors_platform(hass, hass_config.entry_id, platform, dev_ids)
```
```suggestion
        discover_mysensors_platform(hass, entry.entry_id, platform, dev_ids)
```

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
 
 
-async def _gw_start(hass, gateway):
+async def gw_stop(hass, hass_config: ConfigEntry, gateway: BaseAsyncGateway):
```
```suggestion
async def gw_stop(hass, entry: ConfigEntry, gateway: BaseAsyncGateway):
```

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
+async def gw_stop(hass, hass_config: ConfigEntry, gateway: BaseAsyncGateway):
+    """Stop the gateway."""
+    connect_task = hass.data[DOMAIN].get(
+        MYSENSORS_GATEWAY_START_TASK.format(hass_config.entry_id), None
```
```suggestion
        MYSENSORS_GATEWAY_START_TASK.format(entry.entry_id), None
```

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
+
+
+async def _gw_start(
+    hass: HomeAssistantType, hass_config: ConfigEntry, gateway: BaseAsyncGateway
```
```suggestion
    hass: HomeAssistantType, entry: ConfigEntry, gateway: BaseAsyncGateway
```

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
     # Don't use hass.async_create_task to avoid holding up setup indefinitely.
-    connect_task = hass.loop.create_task(gateway.start())
+    hass.data[DOMAIN][
+        MYSENSORS_GATEWAY_START_TASK.format(hass_config.entry_id)
```
```suggestion
        MYSENSORS_GATEWAY_START_TASK.format(entry.entry_id)
```

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, gw_stop)
-    if gateway.device == "mqtt":
+    async def stop_this_gw(_: Event):
+        await gw_stop(hass, hass_config, gateway)
```
```suggestion
        await gw_stop(hass, entry, gateway)
```

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
     gateway_ready = asyncio.Future()
-    gateway_ready_key = MYSENSORS_GATEWAY_READY.format(id(gateway))
-    hass.data[gateway_ready_key] = gateway_ready
+    gateway_ready_key = MYSENSORS_GATEWAY_READY.format(hass_config.entry_id)
```
```suggestion
    gateway_ready_key = MYSENSORS_GATEWAY_READY.format(entry.entry_id)
```

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
         _LOGGER.warning(
             "Gateway %s not ready after %s secs so continuing with setup",
-            gateway.device,
+            hass_config.data[CONF_DEVICE],
```
```suggestion
            entry.data[CONF_DEVICE],
```

MartinHjelmare commented on homeassistant/components/mysensors/handler.py:
```diff
@@ -1,93 +1,104 @@
 
 @HANDLERS.register("set")
-async def handle_set(hass, hass_config, msg):
+async def handle_set(hass, gateway_id: GatewayId, msg: Message) -> None:
```
Please type the functions completely if we add typing.
```suggestion
async def handle_set(hass: HomeAssistantType, gateway_id: GatewayId, msg: Message) -> None:
```

MartinHjelmare commented on homeassistant/components/mysensors/light.py:
```diff
@@ -63,7 +85,7 @@ def white_value(self):
     def assumed_state(self):
         """Return true if unable to access real state of entity."""
-        return self.gateway.optimistic
+        return False
```
This is the default. We can remove the property definition.

MartinHjelmare commented on tests/components/mysensors/test_config_flow.py:
```diff
@@ -0,0 +1,559 @@
+    assert stepuser["type"] == "form"
+    assert not stepuser["errors"]
+
+    actualstep = await hass.config_entries.flow.async_configure(
```
Please use our standard variable names.
```suggestion
    result = await hass.config_entries.flow.async_configure(
```

MartinHjelmare commented on tests/components/mysensors/test_config_flow.py:
```diff
@@ -0,0 +1,559 @@
+async def test_config_serial(hass: HomeAssistantType):
+    """Test configuring a gateway via serial."""
+    step = await get_form(hass, CONF_GATEWAY_TYPE_SERIAL, "gw_serial")
+    flowid = step["flow_id"]
```
Please separate words with underscore in variable names.
```suggestion
    flow_id = step["flow_id"]
```

MartinHjelmare commented on tests/components/mysensors/test_config_flow.py:
```diff
@@ -0,0 +1,559 @@
+
+async def test_config_tcp_invalid_port_too_small(hass: HomeAssistantType):
+    """Test invalid port on a tcp gateway."""
+    await config_invalid(
```
We can use pytest parametrization instead and put the assertion code in one test.

https://docs.pytest.org/en/stable/parametrize.html

MartinHjelmare commented on tests/components/mysensors/test_config_flow.py:
```diff
@@ -0,0 +1,559 @@
+        "homeassistant.components.mysensors.async_setup_entry",
+        return_value=True,
+    ) as mock_setup_entry:
+        result = await async_setup(hass, config)
```
Please don't use integration details directly in the tests. We should set up and start the tests with the core interfaces.

We don't need to test the init module in the config flow tests. That should be done separately in a `test_init.py` module. It's not a requirement for this PR.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

MartinHjelmare commented on tests/components/mysensors/test_config_flow.py:
```diff
@@ -0,0 +1,559 @@
+    with patch("sys.platform", "win32"), patch(
+        "homeassistant.components.mysensors.config_flow.try_connect", return_value=True
+    ):
+        handler = MySensorsConfigFlowHandler()
```
Pleaseuse the core interface to start the test.

functionpointer committed:
MySensors: Rename some stuff

functionpointer committed:
MySensors: use pytest.mark.parametrize

functionpointer committed:
MySensors: Clean up test cases

functionpointer committed:
MySensors: Remove unneeded parameter from devices

functionpointer committed:
Revert "MySensors: add packaging==20.8 for the version parser"

This reverts commit 6b200ee01a3c0eee98176380bdd0b73e5a25b2dd.

functionpointer committed:
Merge remote-tracking branch 'upstream/dev' into mys-config-flow

MartinHjelmare commented on tests/components/mysensors/test_init.py:
```diff
@@ -0,0 +1,175 @@
+        "homeassistant.components.mysensors.async_setup_entry",
+        return_value=True,
+    ) as mock_setup_entry:
+        result = await async_setup(hass, config)
```
Please use `async_setup_component` instead.

functionpointer replied:
Good call! I also took the opportunity to introduce `parametrize` to these tests.

functionpointer committed:
MySensors: Use core interface for testing configuration.yaml import

functionpointer committed:
Merge remote-tracking branch 'upstream/dev' into mys-config-flow

MartinHjelmare commented on tests/components/mysensors/test_init.py:
```diff
@@ -22,154 +24,166 @@
+            },
+            2,
+            True,
+            [],
```
Should this be an empty dict?
```suggestion
            {},
```

functionpointer replied:
Good catch! I didn't notice because i also forgot to change the function name *facepalm*

functionpointer committed:
MySensors: Fix test_init

functionpointer committed:
MySensors: Rename a few variables

MartinHjelmare commented on homeassistant/components/mysensors/__init__.py:
```diff
@@ -112,69 +132,164 @@ def validator(config):
 
 
-async def async_setup(hass, config):
+async def async_setup(hass, config: ConfigType) -> bool:
```
```suggestion
async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
```

MartinHjelmare commented on homeassistant/components/mysensors/__init__.py:
```diff
@@ -112,69 +132,164 @@ def validator(config):
+    hass.data[DOMAIN][MYSENSORS_GATEWAYS][entry.entry_id] = gateway
+
+    async def finish():
+        for platform in SUPPORTED_PLATFORMS_WITH_ENTRY_SUPPORT:
+            await hass.config_entries.async_forward_entry_setup(entry, platform)
```
We can use `asyncio.gather` here. See eg `async_unload_entry`.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,224 @@
+        schema[vol.Optional(CONF_TOPIC_OUT_PREFIX)] = str
+
+        schema = vol.Schema(schema)
+        _LOGGER.debug("mysconfigflow mqtt errors: %s", errors)
```
Do we need this debug log?

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,224 @@
+
+            if gw_type != CONF_GATEWAY_TYPE_MQTT:
+                verification_func = (
+                    is_socket_address
```
We prefer to not use a ternary operator expression if the statement spans more than one line.

MartinHjelmare commented on homeassistant/components/mysensors/gateway.py:
```diff
@@ -154,97 +221,118 @@ def internal_callback(msg):
+async def gw_stop(hass, entry: ConfigEntry, gateway: BaseAsyncGateway):
+    """Stop the gateway."""
+    connect_task = hass.data[DOMAIN].get(
+        MYSENSORS_GATEWAY_START_TASK.format(entry.entry_id), None
```
```suggestion
        MYSENSORS_GATEWAY_START_TASK.format(entry.entry_id)
```

MartinHjelmare commented on homeassistant/components/mysensors/helpers.py:
```diff
@@ -1,78 +1,109 @@
-            {ATTR_DEVICES: new_devices, CONF_NAME: DOMAIN},
-            hass_config,
-        )
+    _LOGGER.debug("discovering platform %s with devIds: %s", platform, new_devices)
```
```suggestion
    _LOGGER.debug("Discovering platform %s with devIds: %s", platform, new_devices)
```

MartinHjelmare commented on homeassistant/components/mysensors/notify.py:
```diff
@@ -1,4 +1,5 @@
 """MySensors notification service."""
+
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/mysensors/strings.json:
```diff
@@ -0,0 +1,22 @@
+    "step": {
+      "user": {
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
```
The strings aren't updated. The English translation file is updated, but that's not the source of truth.

MartinHjelmare reviewed (commented):
Please check if we can enforce typing for the integration now without much trouble.

Modify this line:
https://github.com/home-assistant/core/blob/2ffdc4694aef1482e9bbe814b0af567d5af3af8a/setup.cfg#L45

MartinHjelmare commented:
Mostly some clean up left. See comments above and also some older non resolved comments.

functionpointer committed:
MySensors: cosmetic changes

functionpointer committed:
MySensors: Update strings.json

functionpointer committed:
MySensors: Still more feedback from @MartinHjelmare

functionpointer commented:
> 
> 
> Please check if we can enforce typing for the integration now without much trouble.
> 
> Modify this line:
> 
> https://github.com/home-assistant/core/blob/2ffdc4694aef1482e9bbe814b0af567d5af3af8a/setup.cfg#L45

That's why mypy isnt really doing anything!
Unfortunately it reports 218 errors. Most of them are easy to deal with, but some are a bit more involved.
I found these two annoying ones right at the beginning:
```
homeassistant/components/mysensors/helpers.py:59: error: Argument 1 to "register" of "Registry" has incompatible type "Tuple[str, str]"; expected "str"  [arg-type]
homeassistant/components/mysensors/helpers.py:161: error: "IntEnum" has no attribute "__iter__" (not iterable)  [attr-defined]
```
So it looks like it will take a decent amount of work to fix all these errors. Maybe that can be deferred to a future PR?

MartinHjelmare commented:
Yes, it's ok to wait with completing the typing. :+1:

MartinHjelmare commented on homeassistant/components/mysensors/strings.json:
```diff
@@ -0,0 +1,61 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
This isn't used.

MartinHjelmare replied:
Preferably we should check for the same device among existing entries when validating the user input. We can use `hass.config_entries.async_entries`.

MartinHjelmare replied:
I think this is the last things I'd like us to address before merge.

MartinHjelmare commented on homeassistant/components/mysensors/strings.json:
```diff
@@ -0,0 +1,61 @@
+    "step": {
+      "user": {
+        "data": {
+          "optimistic": "optimistic",
```
This isn't used, since it's not in the user flow.
```suggestion
```

MartinHjelmare commented on homeassistant/components/mysensors/strings.json:
```diff
@@ -0,0 +1,61 @@
+      "user": {
+        "data": {
+          "optimistic": "optimistic",
+          "persistence": "persistence",
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/mysensors/strings.json:
```diff
@@ -0,0 +1,61 @@
+        "data": {
+          "optimistic": "optimistic",
+          "persistence": "persistence",
+          "gateway_type_mqtt": "MQTT"
```
```suggestion
          "gateway_type": "Gateway type"
```

functionpointer committed:
MySensors: Fix typo and remove another unused string

functionpointer committed:
MySensors: Remove unused strings

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

functionpointer committed:
MySensors: More strings.json

functionpointer committed:
MySensors: Fix gateway ready handler

functionpointer committed:
MySensors: Add duplicate detection to config flows

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -58,6 +60,31 @@ def _validate_version(version: str) -> Dict[str, str]:
+    if entry.data[CONF_DEVICE] != user_input[CONF_DEVICE]:
+        return False
+    if gw_type == CONF_GATEWAY_TYPE_TCP:
+        return entry.data.get(CONF_TCP_PORT, 5003) == user_input.get(
```
The entry for tcp gateway should always have a port, right?

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -58,6 +60,31 @@ def _validate_version(version: str) -> Dict[str, str]:
+        )
+    if gw_type == CONF_GATEWAY_TYPE_MQTT:
+        entry_topics = {
+            entry.data.get(CONF_TOPIC_IN_PREFIX),
```
Same for mqtt gateway and topic prefix.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+                        user_input[CONF_PERSISTENCE_FILE]
+                    )
+                    for other_entry in self.hass.config_entries.async_entries(DOMAIN):
+                        if real_persistence_path == self._normalize_persistence_file(
```
The persistence file may be missing from the other entry data, right? We need to handle that.

functionpointer replied:
Indeed we do!

functionpointer committed:
MySensors: Deal with non-existing topics and ports.

Includes unit tests for these cases.

MartinHjelmare reviewed (approved):
Awesome! Please confirm that we're ready to merge.

MartinHjelmare commented on homeassistant/components/mysensors/config_flow.py:
```diff
@@ -0,0 +1,284 @@
+
+            for other_entry in self.hass.config_entries.async_entries(DOMAIN):
+                if _is_same_device(gw_type, user_input, other_entry):
+                    errors["base"] = "already_configured"
```
Please add this error key to the errors in `strings.json`.

If it's an import we should abort the flow. Can we add that?

MartinHjelmare reviewed (changes requested):
Sorry. Found another comment.

functionpointer committed:
Merge remote-tracking branch 'upstream/dev' into mys-config-flow

functionpointer committed:
MySensors: Use awesomeversion instead of packaging

functionpointer committed:
Add string already_configured

functionpointer committed:
MySensors: Abort config flow when config is found to be invalid while importing

functionpointer committed:
MySensors: Copy all error messages to also be abort messages

All error strings may now also be used as an abort reason,
so the strings should be defined

MartinHjelmare commented on homeassistant/components/mysensors/strings.json:
```diff
@@ -0,0 +1,79 @@
+      }
+    },
+    "error": {
+      "already_configured": "Device is already configured",
```
```suggestion
      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]",
```

MartinHjelmare commented on homeassistant/components/mysensors/strings.json:
```diff
@@ -0,0 +1,79 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "Device is already configured",
```
```suggestion
      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]",
```

MartinHjelmare committed:
Use string references

MartinHjelmare reviewed (approved):
Great!
